autofit/non_linear/parallel/sneaky.py

```diff
@@ -349,3 +350,171 @@ def __del__(self):
+            prior_transform_args: Optional[Iterable] = None,
+            prior_transform_kwargs: Optional[dict] = None,
+    ):
+
+        self.fitness_init = fitness
+        self.prior_transform_init = prior_transform
+        self.fitness = fitness_cache
+        self.prior_transform = prior_transform_cache
+        self.fitness_args = fitness_args or ()
+        self.fitness_kwargs = fitness_kwargs or {}
+        self.prior_transform_args = prior_transform_args or ()
+        self.prior_transform_kwargs = prior_transform_kwargs or {}
+        self.processes = processes
+        self.pool = None
+        try:
+            from mpi4py import MPI
+            self.comm = MPI.COMM_WORLD
+        except ModuleNotFoundError:
+            pass
+
+        self._processes = self.comm.size
+        
+        init_args = (
+            self.fitness_init, self.prior_transform_init,
+            self.fitness_args, self.fitness_kwargs,
+            self.prior_transform_args, self.prior_transform_kwargs
+        )
+        initializer(*init_args)
+
+    def check_if_mpi(self):
+
+        return self._processes > 1
+
+    def is_master(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            return_value = self.comm.rank == 0
+        else:
+            return_value = True
+        
+        return return_value
+
+    def wait(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            self.pool.wait()
+        else:
+            pass
+            warnings.warn(
+                "Cannot wait for pool to finish if not using MPI"
+            )
+
+    def __enter__(self):
+        """
+        Activate the mp / mpi pool
+        """
+
+        use_mpi = self.check_if_mpi()
+
+        if use_mpi:
+
+            from schwimmbad import MPIPool
+
+            if self.is_master():
+                logger.info("... using Schwimmbad MPIPool")
+            self.pool = MPIPool(use_dill=True)
+
+        else:
+            if self.is_master():
+                logger.info("... using multiprocessing")
+            self.pool = mp.Pool(
+                processes=self.processes,
+            )
+
+        return self
+
+    def map(
+            self, function: Callable,
+            iterable: Iterable
+    ):
+        """
+        Map a function over an iterable using the map method
+        of the initialized pool.
+
+        Parameters
+        ----------
+        function
+            A function to map
+        iterable
+            An iterable to map over
+
+        """
+        return self.pool.map(function, iterable)
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            self.pool.terminate()
+        except:  # noqa
+            pass
+        try:
+            del (FunctionCache.fitness, FunctionCache.prior_transform,
+                 FunctionCache.fitness_args, FunctionCache.fitness_kwargs,
+                 FunctionCache.prior_transform_args, FunctionCache.prior_transform_kwargs)
```

Comment:
Could these attributes have a default value of None so they did not have to be deleted? Or does the existence of the attributes play some important role in communicating state whilst multi processing?